tikv/mvcc.go

```diff
@@ -704,11 +725,54 @@ func (store *MVCCStore) prewriteMutations(reqCtx *requestCtx, mutations []*kvrpc
 		if err1 != nil {
 			return err1
 		}
-		batch.Prewrite(m.Key, lock)
+		keys = append(keys, m.Key)
+		locks = append(locks, lock)
+	}
+
+	if req.GetTryOnePc() {
+		committed, err := store.tryOnePC(reqCtx, keys, locks, req, minCommitTS,
+			req.OnePcMaxCommitTs)
+		if err != nil {
+			return err
+		}
+		if committed {
+			return nil
+		}
+	}
+
+	batch := store.dbWriter.NewWriteBatch(req.StartVersion, 0, reqCtx.rpcCtx)
+	for i, key := range keys {
+		batch.Prewrite(key, locks[i])
 	}
 	return store.dbWriter.Write(batch)
 }
 
+func (store *MVCCStore) tryOnePC(reqCtx *requestCtx, keys [][]byte, locks []*mvcc.MvccLock,
+	req *kvrpcpb.PrewriteRequest, minCommitTS uint64, maxCommitTS uint64) (bool, error) {
+	if minCommitTS > maxCommitTS {
+		log.Debug("1pc transaction fallbacks due to minCommitTS exceeds maxCommitTS",
+			zap.Uint64("startTS", req.StartVersion),
+			zap.Uint64("minCommitTS", minCommitTS),
+			zap.Uint64("maxCommitTS", maxCommitTS))
+		return false, nil
+	}
+	if minCommitTS < req.StartVersion {
+		log.Fatal("1pc commitTS less than startTS", zap.Uint64("startTS", req.StartVersion), zap.Uint64("minCommitTS", minCommitTS))
+	}
+
+	reqCtx.onePCCommitTS = minCommitTS
+	store.updateLatestTS(minCommitTS)
+	batch := store.dbWriter.NewWriteBatch(req.StartVersion, minCommitTS, reqCtx.rpcCtx)
+	for i, key := range keys {
+		batch.Commit(key, locks[i])
```

Comment:
There's one more problem. Trying to commit a key that is not locked will cause panicking. Looks like there need a special method for `WriteBatch` to implement it.

Reply:
I found that there's a complete raft system which saves the commit message in raftlog, which also need to be changed to support it. @coocood Do you think it's a good idea to change them all? or is there any better idea?

Reply:
We can support raft mode later.

Reply:
We can add a bool in `MvccLockHdr` for onePC.
And do not delete the lock if it is onePC.

Reply:
It sometimes needs to delete the pessimistic lock.

Reply:
We could also check if the lock is pessimistic.